File: Longest Palindrome Substring.py
# Leetcode Problem 5: Longest Palindrome Substring
# Python Code
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) < 2:
            return s

        start = 0
        max_len = 1

        def expand(left, right):
            nonlocal start, max_len

            while left >= 0 and right < len(s) and s[left] == s[right]:
                if right - left + 1 > max_len:
                    start = left
                    max_len = right - left + 1
                left -= 1
                right += 1

        for i in range(len(s)):
            expand(i, i)       # Odd-length palindrome
            expand(i, i + 1)   # Even-length palindrome

        return s[start:start + max_len]

File: Longest Palindrome Substring.java
// Leetcode Problem 5: Longest Palindrome Substring
// Java Code
class Solution {
    private int start = 0;
    private int maxLen = 1;

    public String longestPalindrome(String s) {
        int n = s.length();

        if (n < 2)
            return s;

        for (int i = 0; i < n; i++) {
            expandAroundCenter(s, i, i);       // Odd-length palindrome
            expandAroundCenter(s, i, i + 1);   // Even-length palindrome
        }

        return s.substring(start, start + maxLen);
    }

    private void expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() &&
               s.charAt(left) == s.charAt(right)) {

            if (right - left + 1 > maxLen) {
                start = left;
                maxLen = right - left + 1;
            }

            left--;
            right++;
        }
    }
}